reconcile/utils/oc.py

```diff
@@ -668,6 +682,252 @@ def _run_json(self, cmd, allow_not_found=False):
+                        label_selector=labels)
+                    if item:
+                        items.append(item.to_dict())
+                except NotFoundError:
+                    pass
+            items_list = {'items': items}
+        else:
+            items_list = obj_client.get(
+                namespace=namespace, label_selector=labels).to_dict()
+
+        items = items_list.get('items')
+        if items is None:
+            raise Exception("Expecting items")
+
+        return items
+
+    def get(self, namespace, kind, name=None, allow_not_found=False):
+        k, group_version = self._parse_kind(kind)
+        obj_client = self._get_obj_client(group_version=group_version, kind=k)
+        try:
+            obj = obj_client.get(name=name, namespace=namespace)
+            return obj.to_dict()
+        except NotFoundError as e:
+            if allow_not_found:
+                return {}
+            else:
+                raise StatusCodeError(f"[{self.server}]: {e}")
+
+    def get_all(self, kind, all_namespaces=False):
+        k, group_version = self._parse_kind(kind)
+        obj_client = self._get_obj_client(
+            group_version=group_version, kind=k)
+        try:
+            return obj_client.get().to_dict()
+        except NotFoundError as e:
+            raise StatusCodeError(f"[{self.server}]: {e}")
+
+    @staticmethod
+    def add_group_kind(kind, kgv, new, preferred):
+        updated_kgv = copy.copy(kgv)
+        if kind not in kgv:
+            # this is a new kind so add it
+            updated_kgv[kind] = [new]
+        else:
+            # this kind already exists, so check if this apigroup has
+            # already been added as an option.  If this apigroup/version is the
+            # preferred one, then replace the apigroup/version so that the
+            # preferred apigroup/version is used instead of a non-preferred one
+            group = new.split('/', 1)[0]
+            new_group = True
+            for pos in range(len(kgv[kind])):
+                if group in kgv[kind][pos]:
+                    new_group = False
+                    if preferred:
+                        updated_kgv[kind][pos] = new
+                    break
+
+            if new_group:
+                # this is a new apigroup
+                updated_kgv[kind].append(new)
+        return updated_kgv
+
+
+class OC:
+    client_status = Counter(name='qontract_reconcile_native_client',
+                            documentation='Cluster is using openshift '
+                            'native client',
+                            labelnames=['cluster_name', 'native_client'])
+
+    def __new__(cls, server, token, jh=None, settings=None,
+                init_projects=False, init_api_resources=False,
+                local=False, cluster_name=None):
+        use_native = os.environ.get('USE_NATIVE_CLIENT', False)
+        if use_native:
+            if cluster_name:
```

Comment:
can you remind me in what case the `cluster_name` is expected to be None?

Reply:
IN cases like saas deploy where OC is created just to use the process to generate a template.